lib/pbench/server/database/models/users.py

```diff
@@ -1,58 +1,121 @@
-import datetime
 import enum
 from typing import Optional
 
-from email_validator import validate_email
-from flask_bcrypt import generate_password_hash
-from sqlalchemy import Column, DateTime, Enum, Integer, LargeBinary, String
-from sqlalchemy.orm import relationship, validates
-from sqlalchemy.orm.exc import NoResultFound
+from sqlalchemy import Column, Integer, String
+from sqlalchemy.exc import IntegrityError
+from sqlalchemy.orm import relationship
 
+from pbench.server import JSONOBJECT
 from pbench.server.database.database import Database
 
 
 class Roles(enum.Enum):
     ADMIN = 1
 
 
+class UserError(Exception):
+    """A base class for errors reported by the user class.
+
+    It is never raised directly, but may be used in "except" clauses.
+    """
+
+
+class UserSqlError(UserError):
+    """SQLAlchemy errors reported through User operations.
+
+    The exception will identify the operation being performed and the config
+    key; the cause will specify the original SQLAlchemy exception.
+    """
+
+    def __init__(self, operation: str, params: JSONOBJECT, cause: str):
+        self.operation = operation
+        self.params = params
+        self.cause = cause
+
+    def __str__(self) -> str:
+        return f"Error {self.operation} {self.params!r}: {self.cause}"
+
+
+class UserDuplicate(UserError):
+    """Attempt to commit a duplicate unique value."""
+
+    def __init__(self, user: "User", cause: str):
+        self.user = user
+        self.cause = cause
+
+    def __str__(self) -> str:
+        return f"Duplicate user setting in {self.user.get_json()}: {self.cause}"
+
+
+class UserNullKey(UserError):
+    """Attempt to commit an Audit row with an empty required column."""
+
+    def __init__(self, user: "User", cause: str):
+        self.user = user
+        self.cause = cause
+
+    def __str__(self) -> str:
+        return f"Missing required key in {self.user.get_json()}: {self.cause}"
+
+
 class User(Database.Base):
     """User Model for storing user related details."""
 
     __tablename__ = "users"
 
     id = Column(Integer, primary_key=True, autoincrement=True)
-    username = Column(String(255), unique=True, nullable=False)
-    first_name = Column(String(255), unique=False, nullable=False)
-    last_name = Column(String(255), unique=False, nullable=False)
-    password = Column(LargeBinary(128), nullable=False)
-    registered_on = Column(DateTime, nullable=False, default=datetime.datetime.now())
-    email = Column(String(255), unique=True, nullable=False)
-    role = Column(Enum(Roles), unique=False, nullable=True)
-    auth_tokens = relationship("ActiveTokens", backref="users")
+    oidc_id = Column(String(255), unique=True, nullable=False)
+    username = Column(String(255), unique=True, nullable=True)
+    dataset_metadata = relationship(
+        "Metadata", back_populates="user", cascade="all, delete-orphan"
+    )
+    _roles = Column(String(255), unique=False, nullable=True)
```

Comment:
- I still think we don't want the `id` column:  we'll be tempted to reference it in various ways when we should probably be using `oidc_id` (or `username`).
- I'm not sure that `username` and `_roles` should be nullable:  we are going to want at least a placeholder value for the username for translations from the UUID, and guaranteeing that the roles is at least an "empty list" would be handy on the other end (so we don't have to handle both empty-list _and_ null-entry).

Reply:
Agree on oidc_id and username. But changing the type of `_roles` is tricky because `list` is not a first-class type of sqlalchemy or the SQL database column data types. We have to use something like `sqlalchemy.ext.mutable.MutableList` which I haven't done much experimentation with to feel comfortable. So that's why as a workaround I am keeping it as a `string` but we can always refer to the `roles` attribute (not `_roles`) to get the list of all the roles we encoded as strings in our User table. Just below this line I have a getter and setter for the `roles` attribute.

Reply:
> * I still think we don't want the `id` column:  we'll be tempted to reference it in various ways when we should probably be using `oidc_id` (or `username`)

I think it wouldn't hurt to remove the `id` column for now but please take a look at this [comment](https://github.com/distributed-system-analysis/pbench/pull/3251#discussion_r1117955343) If we have multiple identity providers and if they happen to report the exact same UUID then this would break (extremely rare but still).

Reply:
I'm sorry I wasn't clear, @npalaska:  what I was questioning here was the `nullable=True` attributes, not the representations of the respective values.

I still think that `username` should not be nullable, but I've reconsidered `_roles`:  I think that having a null value represent an empty list of roles actually will end up being no worse than using (e.g.) an empty string (because of the way that `str.split()` works on empty strings).